src/OpenTelemetry.Abstractions/Resources/Resource.cs

```diff
@@ -0,0 +1,120 @@
+﻿// <copyright file="Resource.cs" company="OpenTelemetry Authors">
+// Copyright 2018, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+namespace OpenTelemetry.Resources
+{
+    using System;
+    using System.Collections.Generic;
+    using OpenTelemetry.Utils;
+
+    /// <summary>
+    /// <see cref="Resource"/> represents a resource, which captures identifying information about the entities
+    /// for which signals (stats or traces) are reported.
+    /// </summary>
+    public class Resource
+    {
+        /// <summary>
+        /// Maximum length of the resource type name.
+        /// </summary>
+        private const int MaxResourceTypeNameLength = 255;
+
+        internal Resource(Dictionary<string, string> labels)
+        {
+            this.Labels = ValidateLabels(labels);
+        }
+
+        /// <summary>
+        /// Gets an empty Resource.
+        /// </summary>
+        public static Resource Empty { get; } = new Resource(new Dictionary<string, string>());
+
+        /// <summary>
+        /// Gets the map of tags describing the resource.
+        /// </summary>
+        public Dictionary<string, string> Labels { get; internal set; }
```

Comment:
This may return IEnumerable of tuples. Dictionary may be too much and also makes collection returned mutable

Reply:
Was following Java implementation which used Map.  IEnumerable<Tuple<string, string>> lead to data properties ITem1 and Item2 rather than Key and Value which deviates from Java.  Rather than returning Dictionary I changed to interfaces where possible and use IReadOnlyDictionary to eliminate mutablility.  This still follows Java as closely as possible but eliminates mutable collection being returned.